## Step 1 > import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

#define "root' path 
ICpath = r'C:\Users\joseh\Instacart Basket Analysis'

#import data set orders_products_derived_fields.pkl
ords_prods_merge = pd.read_pickle(os.path.join(ICpath, '02 Data', 'Prepared Data', 'orders_products_derived_fields.pkl'))
ords_prods_merge #verify what has been imported

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0          2539329        1             1          2                  8   
1          2398795        1             2          3                  7   
2           473747        1             3          3                 12   
3          2254736        1             4          4                  7   
4           431534        1             5          4                 15   
...            ...      ...           ...        ...                ...   
32404854   1320836   202557            17          2                 15   
32404855     31526   202557            18          5                 11   
32404856    758936   203436             1          2                  7   
32404857   2745165   203436             2          3                  5   
32404858   3093936   205420             1          4                 14   

          days_since_prior_order  first_order_from_customer  product_id  \
0                            NaN                       True         196   
1                           15.0                      False         196   
2                           21.0                      False         196   
3                           29.0                      False         196   
4                           28.0                      False         196   
...                          ...                        ...         ...   
32404854                     1.0                      False       43553   
32404855                     3.0                      False       43553   
32404856                     NaN                       True       42338   
32404857                    15.0                      False       42338   
32404858                     NaN                       True       28818   

          add_to_cart_order  reordered                   product_name  \
0                         1          0                           Soda   
1                         1          1                           Soda   
2                         1          1                           Soda   
3                         1          1                           Soda   
4                         1          1                           Soda   
...                     ...        ...                            ...   
32404854                  2          1            Orange Energy Shots   
32404855                  2          1            Orange Energy Shots   
32404856                  4          0          Zucchini Chips, Pesto   
32404857                 16          1          Zucchini Chips, Pesto   
32404858                  8          0  Hot Oatmeal Multigrain Raisin   

          aisle_id  department_id  prices        price_label   busiest_day  \
0               77              7     9.0  Mid-range product  Regular busy   
1               77              7     9.0  Mid-range product  Regular busy   
2               77              7     9.0  Mid-range product  Regular busy   
3               77              7     9.0  Mid-range product    Least busy   
4               77              7     9.0  Mid-range product    Least busy   
...            ...            ...     ...                ...           ...   
32404854        64              7     3.7  Low-range product  Regular busy   
32404855        64              7     3.7  Low-range product  Regular busy   
32404856        50             19     6.9  Mid-range product  Regular busy   
32404857        50             19     6.9  Mid-range product  Regular busy   
32404858       130             14    10.3  Mid-range product    Least busy   

          Busiest_Days busiest_period_of_day  
0         regular busy           Most orders  
1         slowest days        Average orders  
2         slowest days           Most orders  
3         slowest days        Average orders  
4         slowest days           Most orders  
...                ...                   ...  
32404854  regular busy           Most orders  
32404855  regular 

## Step 2 > In this Exercise, you learned how to find the aggregated mean of the ```order_number``` column grouped by ```department_id``` for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [2]:
ords_prods_merge.groupby('department_id').agg({'order_number' : ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## Step 3 > Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

### NOTE: for comparison purposes see the result for the first 1M rows in the next cell 

#### Observations
* The subset of the first 1 Million rows displays 8 ```department_id```s VS. the total of 21
* When we take into account the entire dataframe, order_number means is lower in 6 of 8 ```department_id```s (4,7,13,14,16,20)
* The means range from 11.29 to 19.46 in the small sample
* The means range from 15.22 to 22.90 in the full data set

In [4]:
ords_prods_merge[:1000000].groupby('department_id').agg({'order_number' : ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

## Step 4 > Follow the instructions in the Exercise for creating a ```loyalty flag``` for existing customers using the ```transform()``` and ```loc()``` functions.

In [3]:
ords_prods_merge.describe()

order_id       user_id  order_number     order_dow  \
count  3.240486e+07  3.240486e+07  3.240486e+07  3.240486e+07   
mean   1.710745e+06  1.029372e+05  1.714230e+01  2.738867e+00   
std    9.872988e+05  5.946610e+04  1.753532e+01  2.090077e+00   
min    2.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.559470e+05  5.142200e+04  5.000000e+00  1.000000e+00   
50%    1.711049e+06  1.026160e+05  1.100000e+01  3.000000e+00   
75%    2.565499e+06  1.543890e+05  2.400000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  9.900000e+01  6.000000e+00   

       order_hour_of_day  days_since_prior_order    product_id  \
count       3.240486e+07            3.032876e+07  3.240486e+07   
mean        1.342515e+01            1.110408e+01  2.559866e+04   
std         4.246380e+00            8.779064e+00  1.408400e+04   
min         0.000000e+00            0.000000e+00  1.000000e+00   
25%         1.000000e+01            5.000000e+00  1.354400e+04   
50%         1.300000e+01            8.000000e+00  2.530200e+04   
75%         1.600000e+01            1.500000e+01  3.794700e+04   
max         2.300000e+01            3.000000e+01  4.968800e+04   

       add_to_cart_order     reordered      aisle_id  department_id  \
count       3.240486e+07  3.240486e+07  3.240486e+07   3.240486e+07   
mean        8.352547e+00  5.895873e-01  7.119612e+01   9.919792e+00   
std         7.127071e+00  4.919087e-01  3.821139e+01   6.281485e+00   
min         1.000000e+00  0.000000e+00  1.000000e+00   1.000000e+00   
25%         3.000000e+00  0.000000e+00  3.100000e+01   4.000000e+00   
50%         6.000000e+00  1.000000e+00  8.300000e+01   9.000000e+00   
75%         1.100000e+01  1.000000e+00  1.070000e+02   1.600000e+01   
max         1.450000e+02  1.000000e+00  1.340000e+02   2.100000e+01   

             prices  
count  3.240486e+07  
mean   1.198023e+01  
std    4.956554e+02  
min    1.000000e+00  
25%    4.200000e+00  
50%    7.400000e+00  
75%    1.130000e+01  
max    9.999900e+04

In [4]:
df.groupby('department_id').agg({'order_number' : ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [5]:
df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

In [6]:
df.groupby('department_id').order_number.mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

In [7]:
df.groupby('department_id').agg({'order_number' : ['mean','min','max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

In [8]:
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [12]:
#pd.options.display.max_rows = None
ords_prods_merge.head(11)

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0    2539329        1             1          2                  8   
1    2398795        1             2          3                  7   
2     473747        1             3          3                 12   
3    2254736        1             4          4                  7   
4     431534        1             5          4                 15   
5    3367565        1             6          2                  7   
6     550135        1             7          1                  9   
7    3108588        1             8          1                 14   
8    2295261        1             9          1                 16   
9    2550362        1            10          4                  8   
10   2968173       15            15          1                  9   

    days_since_prior_order  first_order_from_customer  product_id  \
0                      NaN                       True         196   
1                     15.0                      False         196   
2                     21.0                      False         196   
3                     29.0                      False         196   
4                     28.0                      False         196   
5                     19.0                      False         196   
6                     20.0                      False         196   
7                     14.0                      False         196   
8                      0.0                      False         196   
9                     30.0                      False         196   
10                     7.0                      False         196   

    add_to_cart_order  reordered product_name  aisle_id  department_id  \
0                   1          0         Soda        77              7   
1                   1          1         Soda        77              7   
2                   1          1         Soda        77              7   
3                   1          1         Soda        77              7   
4                   1          1         Soda        77              7   
5                   1          1         Soda        77              7   
6                   1          1         Soda        77              7   
7                   2          1         Soda        77              7   
8                   4          1         Soda        77              7   
9                   1          1         Soda        77              7   
10                  2          0         Soda        77              7   

    prices        price_label   busiest_day  Busiest_Days  \
0      9.0  Mid-range product  Regular busy  regular busy   
1      9.0  Mid-range product  Regular busy  slowest days   
2      9.0  Mid-range product  Regular busy  slowest days   
3      9.0  Mid-range product    Least busy  slowest days   
4      9.0  Mid-range product    Least busy  slowest days   
5      9.0  Mid-range product  Regular busy  regular busy   
6      9.0  Mid-range product  Regular busy  busiest days   
7      9.0  Mid-range product  Regular busy  busiest days   
8      9.0  Mid-range product  Regular busy  busiest days   
9      9.0  Mid-range product    Least busy  slowest days   
10     9.0  Mid-range product  Regular busy  busiest days   

   busiest_period_of_day  max_order  
0            Most orders         10  
1         Average orders         10  
2            Most orders         10  
3         Average orders         10  
4            Most orders         10  
5         Average orders         10  
6            Most orders         10  
7            Most orders         10  
8            Most orders         10  
9            Most orders         10  
10           Most orders         22

In [13]:
#assigning a new label to express loyalty level
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag' ] = 'Lotalty customer'

In [15]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag' ] = 'New customer'

In [17]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) &
                     (ords_prods_merge['max_order'] > 10), 'loyalty_flag' ] = 'Regular customer'

In [18]:
ords_prods_merge['loyalty_flag'].value_counts()

Regular customer    15876776
Lotalty customer    10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [20]:
#verifying size+frequency of new columns
15876776 + 10284093 + 6243990

32404859

In [21]:
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3